File: 46-permutations/46-permutations.cpp
class Solution {
public:
   vector<vector<int>> res;
    void loop(int index,vector<int > &nums,int nums_size)
    {
        if(index==nums_size-1)
        {
            res.push_back(nums);
            return ;
        }
        for(int i = index; i<nums.size();i++)
        {
            swap(nums[i],nums[index]);
            loop(index+1,nums,nums_size);
            swap(nums[i],nums[index]); //bactrack
        }
    }
    vector<vector<int>> permute(vector<int>& nums) 
    {
        vector<int >ch;
        loop(0,nums,nums.size());    
        return res;
    }
};

File: 46-permutations/46-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
       
        fun(0,nums,ans);
        
        return ans;
    }
    public void swap(int a,int b,int[] nums){
        int temp = nums[a];
       nums[a]=nums[b];
        nums[b]=temp;
    }
    public void fun(int ind,int[] nums, List<List<Integer>> ans){
        if(ind == nums.length){
             List<Integer> res = new ArrayList<>();
        for(int i=0;i<nums.length;i++){
            res.add(nums[i]);
        }
            ans.add(new ArrayList<>(res));
            return;
        }
       
        for(int i=ind;i<nums.length;i++){
            swap(ind,i,nums);
            fun(ind+1,nums,ans);
            swap(ind,i,nums);
        }
    }
}